docs/execution-providers/EP-Context-Design.md

```diff
@@ -0,0 +1,82 @@
+---
+title: EP context design
+description: ONNX Runtime EP context cache feature design
+parent: Execution Providers
+nav_order: 16
+redirect_from: /docs/reference/execution-providers/EP-Context-Design
+---
+
+# OnnxRuntime EP context cache feature design
+{: .no_toc }
+
+## Contents
+{: .no_toc }
+
+* TOC placeholder
+{:toc}
+
+## Background
+
+OnnxRuntime Execution Providers enable users to inference Onnx model on different kinds of hardware accelerators empowered by backend SDKs (like QNN, OpenVINO, Vitis AI, etc). The Execution Providers converts the Onnx model into graph format required by the backend SDK, and compiles it into the format required by the hardware. Specific to NPU world, the converting and compiling process takes a long time to complete, especially for LLM models. The session creation time costs tens of minutes for some cases which impacts the user experience badly.
+To avoid the converting and compiling cost, most of the backend SDKs provide the feature to dump the pre-compiled model into binary file. The pre-compiled model can be loaded by backend SDK directly and executed on the target device. It improves the session creation time greatly by using this way. In order to achieve this, OnnxRuntime defined a contribute Op called EPContext in MS domain.
+
+## EPContext Op Schema
+
+Op domain: com.microsoft
+Node inputs & outputs: variadic
+Domain: com.microsoft
+Atrribures:
+
+|Attributes           |Data type|Description                                                                                               |
+|---------------------|---------|----------------------------------------------------------------------------------------------------------|
+|main_context         |int64    |1 (default): This node points to an EP context content that contains the graph referred to by this node.<br/>0: The node does not point to any EP context content. Expect to get the graph from node with this field is 1.<br/>Qualcomm QNN supports 1 single context contains multiple graphs. The EPContext node with main_context=1 refers to the real context. And the context contains graphs that are referred by other nodes with main_context=0.|
```

Comment:
can we describe this in a way which doesn't need to refer to Qualcomm QNN?
let's discuss.